File: services/coordinator/internal/sas/sas.go
package sas

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/binary"
	"fmt"
	"hash"
	"sort"
	"strings"
)

const (
	Domain   = "phonebeam-sas-v1"
	DomainV2 = "phonebeam-sas-v2"
)

func writePairingTranscript(mac hash.Hash, origin, sid, pid, operatorID string, requestedCaps []string) {
	caps := append([]string{}, requestedCaps...)
	sort.Strings(caps)
	_, _ = mac.Write([]byte("origin=" + origin + "\n"))
	_, _ = mac.Write([]byte("sid=" + sid + "\n"))
	_, _ = mac.Write([]byte("pid=" + pid + "\n"))
	_, _ = mac.Write([]byte("operator=" + operatorID + "\n"))
	_, _ = mac.Write([]byte("caps=" + strings.Join(caps, ",") + "\n"))
}

func digits(sum []byte) string {
	n := binary.BigEndian.Uint32(sum[:4]) % 1_000_000
	return fmt.Sprintf("%03d %03d", n/1000, n%1000)
}

// Display is the M1 pairing SAS. It does not bind WebRTC DTLS fingerprints.
func Display(key []byte, origin, sid, pid, operatorID string, requestedCaps []string) string {
	mac := hmac.New(sha256.New, key)
	_, _ = mac.Write([]byte(Domain + "\n"))
	writePairingTranscript(mac, origin, sid, pid, operatorID, requestedCaps)
	return digits(mac.Sum(nil))
}

// MediaDisplay is the M2 media-authentication SAS.
// Each peer must compute this locally using fingerprints observed on its own PeerConnection
// (local certificate + remote DTLS), never a coordinator-supplied fingerprint pair.
func MediaDisplay(key []byte, origin, sid, pid, operatorID string, requestedCaps []string, androidFP, browserFP string) string {
	mac := hmac.New(sha256.New, key)
	_, _ = mac.Write([]byte(DomainV2 + "\n"))
	writePairingTranscript(mac, origin, sid, pid, operatorID, requestedCaps)
	_, _ = mac.Write([]byte("android_fp=" + NormalizeFingerprint(androidFP) + "\n"))
	_, _ = mac.Write([]byte("browser_fp=" + NormalizeFingerprint(browserFP) + "\n"))
	return digits(mac.Sum(nil))
}

// NormalizeFingerprint lowercases SHA-256 hex and strips colons/spaces.
func NormalizeFingerprint(value string) string {
	var b strings.Builder
	for _, r := range strings.ToLower(strings.TrimSpace(value)) {
		if (r >= '0' && r <= '9') || (r >= 'a' && r <= 'f') {
			b.WriteRune(r)
		}
	}
	return b.String()
}

func ValidSHA256Fingerprint(value string) bool {
	n := NormalizeFingerprint(value)
	return len(n) == 64
}

File: apps/android/app/src/main/java/com/phonebeam/android/webrtc/Sas.kt
package com.phonebeam.android.webrtc

import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

object Sas {
    const val DOMAIN_V1 = "phonebeam-sas-v1"
    const val DOMAIN_V2 = "phonebeam-sas-v2"

    fun normalizeFingerprint(value: String): String {
        return value.lowercase().filter { it in '0'..'9' || it in 'a'..'f' }
    }

    fun validSha256Fingerprint(value: String): Boolean = normalizeFingerprint(value).length == 64

    fun pairingDisplay(
        key: ByteArray,
        origin: String,
        sid: String,
        pid: String,
        operatorId: String,
        requestedCaps: List<String>,
    ): String = digits(hmac(key, DOMAIN_V1, origin, sid, pid, operatorId, requestedCaps, null, null))

    fun mediaDisplay(
        key: ByteArray,
        origin: String,
        sid: String,
        pid: String,
        operatorId: String,
        requestedCaps: List<String>,
        androidFp: String,
        browserFp: String,
    ): String = digits(
        hmac(key, DOMAIN_V2, origin, sid, pid, operatorId, requestedCaps, androidFp, browserFp),
    )

    private fun hmac(
        key: ByteArray,
        domain: String,
        origin: String,
        sid: String,
        pid: String,
        operatorId: String,
        requestedCaps: List<String>,
        androidFp: String?,
        browserFp: String?,
    ): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(key, "HmacSHA256"))
        mac.update("$domain\n".toByteArray())
        mac.update("origin=$origin\n".toByteArray())
        mac.update("sid=$sid\n".toByteArray())
        mac.update("pid=$pid\n".toByteArray())
        mac.update("operator=$operatorId\n".toByteArray())
        mac.update("caps=${requestedCaps.sorted().joinToString(",")}\n".toByteArray())
        if (androidFp != null && browserFp != null) {
            mac.update("android_fp=${normalizeFingerprint(androidFp)}\n".toByteArray())
            mac.update("browser_fp=${normalizeFingerprint(browserFp)}\n".toByteArray())
        }
        return mac.doFinal()
    }

    private fun digits(sum: ByteArray): String {
        val n = ((sum[0].toInt() and 0xff) shl 24 or
            ((sum[1].toInt() and 0xff) shl 16) or
            ((sum[2].toInt() and 0xff) shl 8) or
            (sum[3].toInt() and 0xff)).toLong() and 0xffff_ffffL
        val six = (n % 1_000_000L).toInt()
        return "%03d %03d".format(six / 1000, six % 1000)
    }
}
